cnd/src/storage.rs

```diff
@@ -158,197 +158,228 @@ impl Load<SwapContext> for Storage {
-            .db
-            .do_in_transaction::<_, _, anyhow::Error>(move |conn| {
-                let key = Text(id);
-
-                let swap: Swap = swaps::table
-                    .filter(swaps::local_swap_id.eq(key))
-                    .first(conn)?;
+pub trait LoadTables<A, B> {
+    async fn load_tables(&self, id: LocalSwapId) -> anyhow::Result<Tables<A, B>>;
+}
 
-                let halbit = Halbit::belonging_to(&swap).first::<Halbit>(conn)?;
-                let herc20 = Herc20::belonging_to(&swap).first::<Herc20>(conn)?;
-                let secret_hash = db::SecretHash::belonging_to(&swap)
-                    .first::<db::SecretHash>(conn)
-                    .optional()?;
+/// Convert a protocol table, with associated data, into a swap params object.
+pub trait IntoParams: Sized {
+    type ProtocolTable;
+
+    fn into_params(
+        _: Self::ProtocolTable,
+        _: LocalSwapId,
+        _: RootSeed,
+        _: Role,
+        _: SecretHash,
+    ) -> anyhow::Result<Self>;
+}
 
-                Ok((swap, halbit, herc20, secret_hash))
-            })
-            .await
-            .context(db::Error::SwapNotFound)?;
+/// Data required to load in order to construct spawnable swaps (`spawn::Swap`).
+#[derive(Debug)]
+pub struct Tables<A, B> {
+    pub swap: db::Swap,
+    pub alpha: A, // E.g, Herc20
+    pub beta: B,  // E.g, Hbit
+    pub secret_hash: Option<db::SecretHash>,
+}
 
-        let role = swap.role.0;
-        let secret_hash = match role {
-            Role::Alice => {
-                let swap_seed = self.seed.derive_swap_seed(id);
-                SecretHash::new(swap_seed.derive_secret())
+macro_rules! impl_load_tables {
+    ($alpha:tt, $beta:tt) => {
+        #[async_trait::async_trait]
+        impl LoadTables<$alpha, $beta> for Storage {
+            async fn load_tables(&self, id: LocalSwapId) -> anyhow::Result<Tables<$alpha, $beta>> {
+                use crate::db::schema::swaps;
+
+                let (swap, alpha, beta, secret_hash) = self
+                    .db
+                    .do_in_transaction::<_, _, anyhow::Error>(move |conn| {
+                        let key = Text(id);
+
+                        let swap: Swap = swaps::table
+                            .filter(swaps::local_swap_id.eq(key))
+                            .first(conn)?;
+
+                        let alpha = $alpha::belonging_to(&swap).first::<$alpha>(conn)?;
+                        let beta = $beta::belonging_to(&swap).first::<$beta>(conn)?;
+
+                        let secret_hash = db::SecretHash::belonging_to(&swap)
+                            .first::<db::SecretHash>(conn)
+                            .optional()?;
+
+                        Ok((swap, alpha, beta, secret_hash))
+                    })
+                    .await
+                    .context(db::Error::SwapNotFound)?;
+
+                alpha.assert_side(Side::Alpha)?;
+                beta.assert_side(Side::Beta)?;
```

Comment:
This is the only place where we use `assert_side` right?

Why not just:

```suggestion
	if alpha.side.0 != Side::Alpha || beta.side.0 != Side::Beta {
		anyhow::bail!("...")
	}
```

Reply:
I had that first but it didn't build. Unless I'm mistaken, we don't know the types here so we need to use a trait.

Reply:
This should work. You are inside a macro so the code will just expand. Isn't the field called `side` for all tables?